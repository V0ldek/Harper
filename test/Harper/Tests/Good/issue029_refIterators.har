ref Bag a = {
  data = {
    fst :: a;
    snd :: a;
  }

  ctor :: a -> a -> Bag a;
  ctor fst snd = {
    this.data.fst := fst;
    this.data.snd := snd;

    return this;
  };

  fst :: impure -> a;
  fst = this.data.fst;
  snd :: impure -> a;
  snd = this.data.snd;

  swap :: impure -> ();
  swap = {
    (fst :: a) = this.data.fst;
    this.data.fst := this.data.snd;
    this.data.snd := fst;
  };
};

iterate :: Bag Integer -> RefIterator Integer sideeffect;
iterate b = {
  eval printLn "Before first." ();
  yield b.fst ();
  eval printLn "After first." ();
  yield b.snd ();
  eval printLn "After second." ();
};

main :: sideeffect -> ();
main = {
  (b :: Bag Integer) = Bag.ctor 42 17;
  (iter :: RefIterator Integer sideeffect) = iterate b;

  eval printLn "Before next." ();
  eval printLn (iter.next ()) ();
  eval printLn (iter.current ()) ();
  
  eval printLn "Before mutation." ();
  eval b.swap ();
  eval printLn (iter.current ()) ();
  
  eval printLn "Before next 2." ();
  eval printLn (iter.next ()) ();
  eval printLn (iter.current ()) ();

  eval printLn "Before next 3." ();
  eval printLn (iter.next ()) ();
  eval printLn (iter.current ()) ();
};