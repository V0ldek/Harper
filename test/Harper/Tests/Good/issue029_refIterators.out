Before next.
Before first.
true
42
Before mutation.
42
Before next 2.
After first.
true
42
Before next 3.
After second.
false
42

Execution ended with value: ()